<table align="center" width=100%>
    <tr>
        <td width="10%">
            <img src="in_class.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                    <b> Inclass-Lab  <br>(Day 3)
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## Table of Content

1. **[Large Sample Test](#z)**
2. **[Small Sample Test](#t)**
3. **[Z Proportion Test](#prop)**
4. **[Non-parametric Tests](#non_para)**

**Import the required libraries**

In [1]:
import pandas as pd
import scipy.stats as stats
import numpy as np
import statistics
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import shapiro
from statsmodels.stats import weightstats as stests
import statsmodels.api as sm

### Let's begin with some hands-on practice exercises

<a id = "z"> </a>
## 1. Large Sample Test

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>1. The computer engineer claims that the average time requires to scan a system is less than 2.3 minutes. A graduate student collects the time taken by 32 systems to test the engineer's claim. Check whether the graduate student can use the Z-test for testing the claim with 95% confidence. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    time =  [2.51, 2.1, 2.18, 2.61, 1.9, 1.7, 1.95, 2.1, 1.58, 2.5, 2.19, 1.85, 1.35, 2.4, 2.22, 2.15, 1.82, 1.62, 1.26, 
             2.2, 2.35, 2.5, 2.18, 1.58, 2.24, 1.84, 2.31, 2.24, 1.75, 1.41, 1.57, 2.4]

In [38]:
# H0 : mu >= 2.3
# H1 : mu < 2.3

# given data
time = np.array ([2.51, 2.1, 2.18, 2.61, 1.9, 1.7, 1.95, 2.1, 1.58, 2.5, 2.19, 1.85, 1.35, 2.4, 2.22, 2.15, 1.82, 1.62, 1.26, 
         2.2, 2.35, 2.5, 2.18, 1.58, 2.24, 1.84, 2.31, 2.24, 1.75, 1.41, 1.57, 2.4])


sample_mean = np.mean(time)
mu = 2.3
n = 32
sample_std = np.std(time, ddof = 1)
alpha = 0.05

print('Using shapiro test:')
stat, p_val = shapiro(time)
print('stat:', stat, '\np-value:', p_val)

print('\nUsing z-test:')
# calculate the test statistic
zscore =  (sample_mean - mu) / (sample_std /np.sqrt(n))
print('Z-score:', z_stat)

p_val = stats.norm.cdf(zscore)
print('p-value:', p_val)


# Critical z-value for two-tailed test
zval = round(stats.norm.isf(q = alpha), 2)
print('\nZ-crit =', -zval)

print('\nUsing t-test:')
# use 'ttest_1samp()' to calculate the test statistic and corresponding p-value for one-tailed left-tailed test
t_stat, p_val = stats.ttest_1samp(list(time), popmean = 2.3)
print('t-stat:', t_stat, '\np-value:', p_val/2)

Using shapiro test:
stat: 0.9538483023643494 
p-value: 0.18520332872867584

Using z-test:
Z-score: -4.314055766276573
p-value: 8.014324310930383e-06

Z-crit = -1.64

Using t-test:
t-stat: -4.314055766276573 
p-value: 7.57677677949103e-05


Using shapiro test, p-value is > 0.05. So we can say the data is normally distributed. Further, sample size > 30 and test statistic value from Z-test and t-test are same. So the graduate student can use Z-test for testing the claim.  

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>2. The computer engineer claims that the average time requires to scan a system is less than 2.3 minutes. A graduate student collects the time taken by 32 systems to test the engineer's claim. Use the given data and test the claim using p-value technique with 95% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    time =  [2.51, 2.1, 2.18, 2.61, 1.9, 1.7, 1.95, 2.1, 1.58, 2.5, 2.19, 1.85, 1.35, 2.4, 2.22, 2.15, 1.82, 1.62, 1.26, 
             2.2, 2.35, 2.5, 2.18, 1.58, 2.24, 1.84, 2.31, 2.24, 1.75, 1.41, 1.57, 2.4]

In [42]:
# H0 : mu >= 2.3
# H1 : mu < 2.3

# given data
time = np.array ([2.51, 2.1, 2.18, 2.61, 1.9, 1.7, 1.95, 2.1, 1.58, 2.5, 2.19, 1.85, 1.35, 2.4, 2.22, 2.15, 1.82, 1.62, 1.26, 
         2.2, 2.35, 2.5, 2.18, 1.58, 2.24, 1.84, 2.31, 2.24, 1.75, 1.41, 1.57, 2.4])


sample_mean = np.mean(time)
mu = 2.3
n = 32
sample_std = np.std(time, ddof = 1)
alpha = 0.05

print('\nUsing t-test:')
# use 'ttest_1samp()' to calculate the test statistic and corresponding p-value for one-tailed left-tailed test
t_stat, p_val = stats.ttest_1samp(list(time), popmean = 2.3)
print('t-stat:', t_stat, '\np-value:', p_val/2)

# Critical t-value for two-tailed test
t_crit = round(stats.t.isf(q = 0.05, df = 31), 2)
print('\nt-crit =', -t_crit)


Using t-test:
t-stat: -4.314055766276573 
p-value: 7.57677677949103e-05

t-crit = -1.7


* Here the p-value is lesser than alpha (0.05)
* The t-statistic is lesser than -1.7 (t-crit) i.e. t-statistic falls in rejection region of H0
* Thus we reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>3. The sales manager at a car showroom claims that the cars in the showroom have average mileage per gallon (mpg) more than 18 mpg. Use the given dataset to test the manager's claim using critical value method at 90% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the dataset given in the CSV file `cars.csv`

In [14]:
# H0 : mu <= 18
# H1 : mu > 18

cars = pd.read_csv('S6_cars.csv')
mpg = cars['mpg']
cars.head()

name   mpg  cyl   disp   hp  drat     wt   qsec  vs am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0  M     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0  M     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1  M     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1  A     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0  A     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

In [19]:
# perform Shapiro-Wilk test to test the normality
stat, p_value = shapiro(mpg)

# print the test statistic and corresponding p-value 
print('Test statistic:', stat)
print('p-Value:', p_value)

Test statistic: 0.9475650191307068
p-Value: 0.12288373708724976


* From the above result, we can see that the p-value is greater than 0.1, thus we can say that the data is normally distributed.

In [18]:
# calculate the z-value for 90% of confidence level
z_crit = round(stats.norm.isf(q = 0.1), 2)

print('Critical value for one-tailed Z-test:', z_crit)

# Use one-tailed, right-tailed Z-test
z_score, pval = stests.ztest(x1 = mpg, value = 18, alternative = 'larger')

# print the test statistic and corresponding p-value
print("\nZ-score: ", z_score)
print("p-value: ", pval)

Critical value for one-tailed Z-test: 1.28

Z-score:  1.962247030028023
p-value:  0.02486686545272649


* Here the z-score is greater than 1.28 (z-crit) i.e. z-score falls in rejection region of H0
* The p-value is also lesser than alpha (0.1)
* Thus we reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>4. The company MangoFun that produces mango juice claims that the protein content in their juice is more than the juice produced by their competitor company FruitMix. Protein content in 50 boxes of MangoFun juice and 80 boxes of FruitMix juice is collected from a normal population which have an average protein content as 0.4 g and 0.35 g. The two samples have a standard deviation of 0.08 g for MangoFun juice and 0.05 g for FruitMix juice. Test the claim using a critical value method at a 1% level of significance.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [7]:
# H0 : mu_mangofun - mu_fruitmix <= 0
# H1 : mu_mangofun - mu_fruitmix > 0 

# Population standard deviations are not known.So use Two sample Unpaired, right-tailed t-test

# given data
n_mangofun = 50
n_fruitmix = 80
df_juice = n_mangofun + n_fruitmix - 2

xbar_mangofun = 0.4
xbar_fruitmix = 0.35

samplestd_mangofun = 0.08
samplestd_fruitmix = 0.05

# calculate pooled standard deviation
s = np.sqrt((((n_mangofun-1)*samplestd_mangofun**2) +  ((n_fruitmix-1)*samplestd_fruitmix**2)) / df_juice)

# calculate the test statistic 
t_stat = (xbar_mangofun - xbar_fruitmix) / (s * np.sqrt(1/n_mangofun + 1/n_fruitmix))
print('Test Statistic:', t_stat)

# calculate p-value
p_val = stats.t.sf(t_stat, df = df_juice)
print('p-value:', p_val)

# Critical t-value for right-tailed test
t_crit = round(stats.t.isf(q = 0.01, df = df_juice), 2)
print('\nt-crit =', t_crit)

Test Statistic: 4.389149444162152
p-value: 1.1785231868001904e-05

t-crit = 2.36


* Here the t-statistic is greater than 2.36 (t-crit) i.e. t-statistic falls in rejection region of H0
* The p-value is also lesser than alpha (0.01)
* Thus we reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>5. The technician wants to test whether the price of a phablet is not the same as the price of tablet. For the study, the technician collects the data of 50 gadgets of each type. Check whether the data satisfies the conditions to use the two sample Z-test. (Use 5% level of significance)</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Consider the dataset given in the CSV file `electronic.csv`. 

In [30]:
# H0 : mu_phablet - mu_tablet = 0
# H1 : mu_phablet - mu_tablet != 0

electronic = pd.read_csv('S6_electronic.csv')
electronic[electronic['Type'] == 'Phablet'].head()

Screensize (inch)  Price ($)  Weight (g)     Type
0                6.5        245         252  Phablet
1                6.3        210         252  Phablet
2                6.1        224         234  Phablet
3                6.0        217         212  Phablet
4                6.4        252         252  Phablet

In [31]:
electronic[electronic['Type'] == 'Tablet'].head()

Screensize (inch)  Price ($)  Weight (g)    Type
50                9.8        224         517  Tablet
51                9.2        224         495  Tablet
52                9.7        217         539  Tablet
53                8.3        161         440  Tablet
54                9.3        196         506  Tablet

In [39]:
# price of phablets, number of phablets, mean price of phablets, standard deviation of phablets
price_phablet = electronic[(electronic['Type'] == 'Phablet')]['Price ($)']
n_phablet = len(price_phablet)
mean_phablet = np.mean(price_phablet)
std_phablet = statistics.stdev(price_phablet)

# price of tablets, number of tablets, mean price of tablets, standard deviation of tablets
price_tablet = electronic[(electronic['Type'] == 'Tablet')]['Price ($)']
n_tablet = len(price_tablet)
mean_tablet = np.mean(price_tablet)
std_tablet = statistics.stdev(price_tablet)

print('n_phablet:', n_phablet)
print('mean_phablet:', mean_phablet)
print('std_phablet:', std_phablet)

print('\nn_tablet:', n_tablet)
print('mean_phablet:', mean_tablet)
print('std_phablet:', std_tablet)

n_phablet: 50
mean_phablet: 239.26
std_phablet: 26.671707856828366

n_tablet: 50
mean_phablet: 193.9
std_phablet: 21.965882636488796


In [40]:
# perform Shapiro-Wilk test
stat, p_value = shapiro(electronic['Price ($)'])

print('Test statistic:', stat)
print('P-Value:', p_value)

Test statistic: 0.9888564348220825
P-Value: 0.5733178853988647


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the prices of phablets and tablets are normally distributed.


Let us check the equality of variances.

In [41]:
# perform Levene's test
stat, p_value = stats.levene(price_phablet, price_tablet)

print('Test statistic:', stat)
print('P-Value:', p_value)

Test statistic: 0.6635459332943239
P-Value: 0.4172859681296204


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the population variances are equal.

* Therefore, the data satisfies the conditions to use the two sample Z-test

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>6. The technician claims that the price of phablet is not the same as the price of tablet. For the study, the technician collects the data of 50 gadgets of each type. Test the technician's claim using the 95% confidence interval.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Consider the dataset given in the CSV file `electronic.csv`. 

In [44]:
# H0 : mu_phablet - mu_tablet = 0
# H1 : mu_phablet - mu_tablet != 0

# alpha = 0.05

# calculate the z-value for 95% of confidence level
z_crit = round(stats.norm.isf(q = 0.05/2), 2)
print('Critical value for two-tailed Z-test:', z_crit, 'and', -z_crit)

# Use two-tailed Z-test
z_score, pval = stests.ztest(x1 = price_phablet, x2 = price_tablet, value = 0, alternative = 'two-sided')

# print the test statistic and corresponding p-value
print("\nZ-score: ", z_score)
print("p-value: ", pval*2)

Critical value for two-tailed Z-test: 1.96 and -1.96

Z-score:  9.282772555558113
p-value:  3.302494646903457e-20


* Here the z-score is greater than 1.96 (z-crit) i.e. z-score falls in rejection region of H0
* The p-value is also lesser than alpha (0.05)
* Thus we reject the null hypothesis

<a id="t"></a>
## 2. Small Sample Test

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>7. The quality assurance department has collected 15 packets of potato chips. The department wants to test the average weight of the packets is more than 30 g. Check whether they can use the one sample t-test for the population mean with 95% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    pack_wt = [26.8, 29.6, 27.8, 31.2, 30.9, 27.1, 28, 28.6, 29.4, 29.3, 31.5, 32.4, 29.7, 28.1, 31.9]

In [75]:
# given data
pack_wt = [26.8, 29.6, 27.8, 31.2, 30.9, 27.1, 28, 28.6, 29.4, 29.3, 31.5, 32.4, 29.7, 28.1, 31.9]

# perform Shapiro-Wilk test to test the normality
stat, p_value = shapiro(pack_wt)

print('Test statistic:', stat)
print('p-value:', p_value)

Test statistic: 0.9538426995277405
p-value: 0.5868644714355469


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the data is normally distributed.

* Therefore, we can use the one sample t-test

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>8. Use the weight of the potato chips packets given in the previous question and test whether the average weight of packet is more than 30 g. using p-value method. (Use 10% level of significance)</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    pack_wt = [26.8, 29.6, 27.8, 31.2, 30.9, 27.1, 28, 28.6, 29.4, 29.3, 31.5, 32.4, 29.7, 28.1, 31.9]

In [50]:
# given data
pack_wt = [26.8, 29.6, 27.8, 31.2, 30.9, 27.1, 28, 28.6, 29.4, 29.3, 31.5, 32.4, 29.7, 28.1, 31.9]

# H0 : mu <= 30
# H1 : mu > 30

# alpha = 0.1
# df = 15-1

# calculate the t-value for 90% of confidence level
t_crit = round(stats.t.isf(q = 0.1, df = 14), 2)
print('t-crit =', t_crit)

print('\nUsing t-test:')
# use 'ttest_1samp()' to calculate the test statistic and corresponding p-value for one-tailed right-tailed test
t_stat, p_val = stats.ttest_1samp(pack_wt, popmean = 30)
print('t-stat:', t_stat, '\np-value:', p_val/2)

t-crit = 1.35

Using t-test:
t-stat: -1.1204731591740873 
p-value: 0.14068288127820355


In [52]:
# Alternate method
sample_mean = np.mean(pack_wt)
mu = 30
n = 15
sample_std = np.std(pack_wt, ddof = 1)
alpha = 0.1

# calculate t-score using formulae
t_stat = (sample_mean - mu) / (sample_std / np.sqrt(n))
print('t-stat:', t_stat)

# calculate p-value
p_val = stats.t.cdf(t_stat, df = 14)
print('p-value:', p_val)

t-stat: -1.1204731591740873
p-value: 0.14068288127820355


* Here the p-value is greater than the alpha (0.1)
* The t-statistic is also lesser than the t-crit value (1.35) i.e. t-statistic falls in acceptance region of H0
* Therefore, we fail to reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>9. The orthopaedic surgeon states that a T-score for females older than 30 years is less than -1.2 which indicates the low bone density. To test the claim a sample of 10 women was selected and the average T-score was found to be -1.34 with a standard deviation of 0.8. Test the surgeon's claim using a critical value method with 95% confidence. (Assume the normality of the T-score). </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [54]:
# H0 : mu >= -1.2
# H1 : mu < -1.2

# given data
sample_mean = -1.34
n = 10
s = 0.8
mu = -1.2
alpha = 0.05

# calculate the t-value for 95% of confidence level
t_crit = stats.t.isf(q = 0.05, df = 9)
print('t-crit:', -t_crit)

print('\nUsing left-tailed t-test:')
# calculate t-score using formulae
t_stat = (sample_mean - mu) / (s / np.sqrt(n))
print('t-stat:', t_stat)

# calculate p-value
p_val = stats.t.cdf(t_stat, df = 9)
print('p-value:', p_val)

t-crit: -1.8331129326536337

Using left-tailed t-test:
t-stat: -0.5533985905294669
p-value: 0.29673469801808217


* The t-statistic is greater than the t-crit (-1.83)
* The t-statistic falls to the right of t-crit i.e. t-statistic falls in acceptance region of H0
* Also the p-value is greater than alpha (0.05)
* Therefore we fail to reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>10. The newspaper agency wants to test whether Harry takes less time to deliver the newspapers than his colleague Ron. The manager has collected time (in minutes) taken by Harry and Ron for 7 days. Is the given dataset unpaired? If yes, check whether the manager at newspaper agency can use the two sample t-test for unpaired data with 99% confidence.  </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    harry = [18.5, 17.4, 19.2, 16, 15.8, 13.4, 19.5]
    ron = [19.7, 18.6, 21.3, 17.5, 23.8, 20.7, 21]

In [73]:
# given data
# dataset is unpaired
harry = [18.5, 17.4, 19.2, 16, 15.8, 13.4, 19.5]
ron = [19.7, 18.6, 21.3, 17.5, 23.8, 20.7, 21]

The two sample t-test is used to compare the means of two independent populations. This test assumes that the populations are normally distributed from which the samples are taken.

In [76]:
# perform Shapiro-Wilk test to test the normality
stat, p_value = shapiro(harry + ron)

print('Test statistic:', stat)
print('p-value:', p_value)

Test statistic: 0.986557126045227
p-value: 0.9969169497489929


From the above result, we can see that the p-value is greater than 0.01, thus we can say that the data is normally distributed.


Let us check the equality of variances.

In [77]:
# perform Levene's test
stat, p_value = stats.levene(harry, ron)

print('Test statistic:', stat)
print('p-value:', p_value)

Test statistic: 0.12793271359008374
p-value: 0.7267972480247995


From the above result, we can see that the p-value is greater than 0.01, thus we can say that the population variances are equal.

* Thus, two-sample t-test for unpaired data can be used

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>11. The newspaper agency wants to test whether Harry takes less time to deliver the newspapers than his colleague Ron. The manager has collected time (in minutes) taken by Harry and Ron for 7 days. Use the given data and test the hypothesis for population mean time using a critical value method and p-value criteria with 90% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    harry = [18.5, 17.4, 19.2, 16, 15.8, 13.4, 19.5]
    ron = [19.7, 18.6, 21.3, 17.5, 23.8, 20.7, 21]

In [79]:
# H0 : mu_harry - mu_ron >= 0
# H1 : mu_harry - mu_ron < 0 

# given data
harry = [18.5, 17.4, 19.2, 16, 15.8, 13.4, 19.5]
ron = [19.7, 18.6, 21.3, 17.5, 23.8, 20.7, 21]

# calculate the t-value for 90% of confidence level
t_crit = stats.t.isf(q = 0.1, df = 6)
print('t-crit:', -t_crit)

print('\nUsing left-tailed t-test:')
# use 'ttest_ind()' to calculate the test statistic and corresponding p-value for 2 sample test
t_stat, p_val = stats.ttest_ind(a = harry, b = ron)
print('t-stat:', t_stat, '\np-value:', p_val/2)

t-crit: -1.4397557472577691

Using left-tailed t-test:
t-stat: -2.8788582527177526 
p-value: 0.00692997580305821


* The t-statistic is lesser than t-crit (-1.43) i.e. t-statistic falls in rejection region of H0
* The p-value is also less than alpha (0.1)
* Thus we reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                        <b>12. The pharmaceutical company had organized a program to introduce its new drug to lower the sugar level. They recorded the fasting sugar (in mg/dl) of 25 diabetic people. Those people undergo a medication of the new drug for 10 days and again took a fasting sugar level test. The company claim that the sugar level decreases due to its new drug. Test the claim using p-value technique with 95% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given in the file `sugar_level.xlsx`

In [127]:
sugar_level = pd.read_excel('S6_sugar_level.xlsx')
sugar_level.head()

sugar_before  sugar_after
0           135          132
1           132          136
2           142          139
3           154          151
4           198          192

In [131]:
# A paired t-test is used to compare the mean of the population for two dependent samples
# The median of the difference between the two paired samples is denoted by mu_d
# mu_d = Mean(after) - Mean(before)

# H0 : mu_d >= 0
# H1 : mu_d < 0

# calculate the difference between the sugar levels
diff_sugar = sugar_level['sugar_after'] - sugar_level['sugar_before']

# calculate mean difference
mean_diff = np.mean(diff_sugar)

# calculate standard deviation of difference
std_diff = statistics.stdev(diff_sugar)

# sample size
n = len(sugar_level)

print('Mean difference in sugar level:', mean_diff)
print('Std deviation of difference in sugar level:', std_diff)
print('Degrees of freedom:', n-1)

Mean difference in sugar level: -5.08
Std deviation of difference in sugar level: 4.821134029803915
Degrees of freedom: 24


In [133]:
# perform Shapiro-Wilk test to test the normality for sugar level after drug treatment
stat, p_value = shapiro(sugar_level['sugar_before'])

# print the test statistic and corresponding p-value
print('Test statistic:', stat)
print('p-value:', p_value)

Test statistic: 0.9284526109695435
p-value: 0.08004998415708542


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the sugar level before drug treatment are normally distributed.

In [134]:
# perform Shapiro-Wilk test to test the normality for sugar level after drug treatment
stat, p_value = shapiro(sugar_level['sugar_after'])

# print the test statistic and corresponding p-value
print('Test statistic:', stat)
print('p-value:', p_value)

Test statistic: 0.922091543674469
p-value: 0.057184845209121704


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the sugar level after drug treatment are normally distributed.

In [138]:
# H0 : mu_d >= 0
# H1 : mu_d < 0

# calculate the t-value for 90% of confidence level
t_crit = round(stats.t.isf(q = 0.05, df = 24), 2)
print('t-crit:', -t_crit)

print('\nUsing left-tailed t-test:')
# use 'ttest_rel()' to calculate the t-statistic and corresponding p-value for paired samples
t_stat, p_val = stats.ttest_rel(sugar_level['sugar_after'], sugar_level['sugar_before'])
print('t-stat:', t_stat, '\np-value:', p_val/2)

t-crit: -1.71

Using left-tailed t-test:
t-stat: -5.268469999584946 
p-value: 1.0544706972911436e-05


* The p-value is lesser than alpha value (0.05)
* Also, the t-statistic is lesser than the t-crit value (-1.71)
* t-statistic lies to the left of t-crit value i.e. it falls in the rejection region of H0
* This we reject the null hypothesis

<a id= "prop"></a>
## 3. Z Proportion Test

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>13. The physics department claims that it received 14% of fraud applications this year. The university head wants to test whether the percentage is different than what the department claims. The sample of 250 applications is selected out of which 38 applications are found to fraud. Test the claim with a 95% confidence interval.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [92]:
# H0 : p = 0.14
# H1 : p != 0.14

# sample size
n = 250

# number of fraud
x = 38 

# sample proportion
p_samp = x/n

# hypothesized proportion
p_hypo = 0.14

# calculate the z-value for 95% of confidence level
z_crit = round(stats.norm.isf(q = 0.05/2), 2)
print('Critical value for two-tailed Z-test:', z_crit, 'and', -z_crit)

print('\nUsing two-tailed Z-test:')
# calculate test statistic value for 1 sample proportion test
z_score = (p_samp - p_hypo) / np.sqrt((p_hypo * (1 - p_hypo)) / n)
print('Z-score', z_score)

# calculate p-value
p_val = stats.norm.sf(z_score)
print('p-value', p_val*2)

# calculate the 95% confidence interval
interval = stats.norm.interval(0.95, loc = p_samp, scale = np.sqrt(p_samp * (1 - p_samp)/n))
print('\n95% confidence interval for population proportion is', interval)

Critical value for two-tailed Z-test: 1.96 and -1.96

Using two-tailed Z-test:
Z-score 0.5468119624323661
p-value 0.5845079238217354

95% confidence interval for population proportion is (0.10749612708304186, 0.19650387291695814)


* The 95% confidence interval contains the value in the null hypothesis (0.14)
* The z-score is lesser than z-crit (1.96) i.e. z-score falls in acceptance region of H0
* The p-value is greater than alpha (0.05) 
* Thus we fail to reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>14. The store manager claims that he received more 20% of defective plastic boxes in his previous order. Test the manager's claim using a critical value method. From a sample of 120 boxes, 42 were found to be defective. Test the claim with 90% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [94]:
# H0 : p <= 0.2
# H1 : p > 0.2

# sample size
n = 120

# number of defective boxes
x = 42

# sample proportion
p_samp = x/n

# hypothesized proportion
p_hypo = 0.2

# calculate the z-value for 90% of confidence level
z_crit = round(stats.norm.isf(q = 0.1), 2)
print('Critical value for right-tailed Z-test:', z_crit)

print('\nUsing right-tailed Z-test:')
# calculate test statistic value for 1 sample proportion test
z_score = (p_samp - p_hypo) / np.sqrt((p_hypo * (1 - p_hypo)) / n)
print('Z-score', z_score)

# calculate p-value
p_val = stats.norm.sf(z_score)
print('p-value', p_val)

# calculate the 90% confidence interval
interval = stats.norm.interval(0.90, loc = p_samp, scale = np.sqrt(p_samp * (1 - p_samp)/n))
print('\n90% confidence interval for population proportion is', interval)

Critical value for two-tailed Z-test: 1.28

Using right-tailed Z-test:
Z-score 4.107919181288744
p-value 1.996198738450681e-05

90% confidence interval for population proportion is (0.2783811514679558, 0.4216188485320441)


* The 90% confidence interval does not contain the value in the null hypothesis (0.2)
* The z-score is greater than z-crit (1.28) i.e. z-score falls in rejection region of H0
* The p-value is lesser than alpha (0.1)
* Thus we reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                        <b>15. The HR head of a company claims that the company supports the woman empowerment and the proportion of female employees is same in the New York and Oneonta branches. The women empowerment cell wants to check whether the proportion is different for the two branches. The sample of 150 employees is selected from New York branch out of which 53 are females and a sample of 170 employees is selected from Oneonta branch out of which 76 are females. Use the p-value technique with 95% confidence. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [100]:
# H0 : p_ny = p_oo
# H1 : p_ny != p_oo

# sample size New York branch
n_ny = 150

# number of females New York branch
x_ny = 53

# sample size Oneonta branch
n_oo = 170

# number of females Oneonta branch
x_oo = 76

# calculate the z-value for 95% of confidence level
z_crit = round(stats.norm.isf(q = 0.05/2), 2)
print('Critical value for two-tailed Z-test:', z_crit, 'and', -z_crit)

# calculate test statistic value for two sample proportion test
z_score, p_val = sm.stats.proportions_ztest(count = np.array([x_ny, x_oo]), 
                                             nobs = np.array([n_ny, n_oo]),  
                                             alternative = 'two-sided')

print('\nZ-score:', z_score)
print('p-value:', p_val)

Critical value for two-tailed Z-test: 1.96 and -1.96

Z-score: -1.705655272664087
p-value: 0.08807228185564836


* The p-value is greater than alpha (0.05) 
* Also, the z-score is greater than z-crit (-1.96) i.e. it falls in acceptance region of H0
* Thus we fail to reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>16. The education department claims that the proportion of students failed in Mathematics is more than the proportion of students failed in English. To test the claim a sample of 200 students enrolled for Mathematics was considered out of which 73 students failed in the exam and a sample of 150 students enrolled for English was considered out of which 53 students failed in the exam. Use the critical value method to test the department's claim with 90% confidence. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [101]:
# H0 : p_math <= p_english
# H1 : p_math > p_english

# sample size Math
n_M = 200

# number of students failed in Math
x_M =73

# sample size English
n_E = 150

# number of students failed in English
x_E = 53

# calculate the z-value for 90% of confidence level
z_crit = round(stats.norm.isf(q = 0.1), 2)
print('Critical value for two-tailed Z-test:', z_crit)

# calculate test statistic value for two sample proportion test
z_score, p_val = sm.stats.proportions_ztest(count = np.array([x_M, x_E]), 
                                             nobs = np.array([n_M, n_E]),  
                                             alternative = 'larger')

print('\nZ-score:', z_score)
print('p-value:', p_val)

Critical value for two-tailed Z-test: 1.28

Z-score: 0.22502571869471721
p-value: 0.41097963333926457


* The z-score is lesser than z-crit (1.28) i.e. z-score falls in acceptance region of H0
* Also, the p-value is greater than alpha (0.1) 
* Thus we fail to reject the null hypothesis

<a id= "non_para"></a>
## 4. Non-parametric Tests

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>17. The car dealer claims that the median horsepower (hp) of the cars in his showroom is greater than 100 watts. Test the dealer's claim with 95% confidence.</b>
                </font> 
            </div>
        </td>
    </tr>
</table>

Use the dataset given in the CSV file `cars.csv`

In [108]:
# H0 : mu <= 100
# H1 : M > 100

# Wilcoxon signed rank test (One-sample test) is used to compare the median (M) of a sample to a specific value (M0).

cars = pd.read_csv('S6_cars.csv')
cars.head()

name   mpg  cyl   disp   hp  drat     wt   qsec  vs am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0  M     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0  M     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1  M     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1  A     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0  A     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

In [109]:
# given horsepower
hp = cars['hp']

# hypothesized median
M0 = 100

# calculate the difference between horsepower and M0
diff = hp - M0

# perform wilcoxon signed rank test
test_stat, p_value = stats.wilcoxon(x = diff, alternative = 'greater')

# print the test statistic and corresponding p-value
print('Test statistic:', test_stat)
print('p-value:', p_value)

Test statistic: 440.0
p-value: 0.0004968423513640346


* Here the p-value is lesser than alpha (0.05)
* Thus we reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>18. The customer claims that the median weight of watermelon in HealthyFood store is different than 2.5 kg. 8 watermelons are selected as a sample and the weight is recorded. Test the customer's claim at 90% confidence using p-value technique. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Given data:

        watermelon_wt = [1.81, 2.3, 1.8, 2.65, 3.9, 3.7, 1.9, 2.1]

In [104]:
# H0 : M = 2.5
# H1 : M != 2.5

# Wilcoxon signed rank test (One-sample test) is used to compare the median (M) of a sample to a specific value (M0).

# given watermelon weights
watermelon_wt = np.array([1.81, 2.3, 1.8, 2.65, 3.9, 3.7, 1.9, 2.1])

# hypothesized median
M0 = 2.5

# calculate the difference between weights and M0
diff = watermelon_wt - M0

# perform wilcoxon signed rank test
test_stat, p_value = stats.wilcoxon(x = diff, alternative = 'two-sided')

# print the test statistic and corresponding p-value
print('Test statistic:', test_stat)
print('p-value:', p_value)

Test statistic: 16.0
p-value: 0.84375


* Here the p-value is greater than alpha (0.1)
* Thus we fail to reject the null hypothesis 

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>19. The gym trainer claims that the new aerobic exercise has reduced the weight of the customers. The weight in the last month before the aerobic exercise and weight in this month after the aerobic exercise is recorded for 6 people. Test the trainer's claim with 95% confidence.  </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Given data:

        wt_before = [85, 74, 63.5, 69.4, 71.6, 65]
        wt_after = [82, 71, 64, 65.2, 67.8, 64.7]

In [110]:
# Wilcoxon signed rank test (Two-sample paired test) can be used to compare medians of paired data
# The median of the difference between the two paired samples is denoted by Md
# Md = Median(before) - Median(after)

# H0 : Md <= 0
# H1 : Md > 0

# given data
wt_before = [85, 74, 63.5, 69.4, 71.6, 65]
wt_after = [82, 71, 64, 65.2, 67.8, 64.7]

# perform wilcoxon signed rank test for paired data
test_stat, p_value = stats.wilcoxon(x = wt_before, y = wt_after, alternative = 'greater')

# print the test statistic and corresponding p-value
print('Test statistic:', test_stat)
print('p-value:', p_value)

Test statistic: 19.0
p-value: 0.046875


* The p-value is lesser than alpha (0.05) 
* Thus we reject the null hypothesis

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>20. Consider the cars dataset given below. Check whether the distribution of weight of the manual transmission cars is different than the distribution of weight of the automatic transmission cars. Use a 1% level of significance to check the noramlity of the data. And use the appropriate test with 90% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the dataset given in the CSV file `cars.csv`

In [111]:
# H0 : F_manual = F_auto
# H1 : F_manual != F_auto

# The Mann-Whitney U test is a non-parametric test that compares the 
# distributions of independent populations. 
#This test can be used as a non-parametric alternative for the unpaired t-test. 

cars = pd.read_csv('S6_cars.csv')
cars.head()

name   mpg  cyl   disp   hp  drat     wt   qsec  vs am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0  M     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0  M     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1  M     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1  A     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0  A     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

In [122]:
# weight of manual transmission cars
wt_manual = cars[cars['am'] == 'M']['wt']

# perform Shapiro-Wilk test to test the normality
stat, p_value = shapiro(wt_manual)

# print the corresponding p-value 
print('p-value:', p_value)

p-value: 0.9091034531593323


From the above result, we can see that the p-value is greater than 0.01, thus we can say that the weight of the manual transmission cars are normally distributed.

In [123]:
# weight of automatic transmission cars
wt_automatic = cars[cars['am'] == 'A']['wt']

# perform Shapiro-Wilk test to test the normality
stat, p_value = shapiro(wt_automatic)

# print the corresponding p-value 
print('p-value:', p_value)

p-value: 0.003307320410385728


From the above result, we can see that the p-value is lesser than 0.01, thus we can say that the weight of the automatic transmission cars are not normally distributed.

In [124]:
# compute test statistic and corresponding p-value for two-tailed test
test_stat, p_value = stats.mannwhitneyu(wt_manual, wt_automatic, alternative = 'two-sided')

# print the test statistic and corresponding p-value
print('Test statistic:', test_stat)
print('p-value:', p_value)

Test statistic: 16.5
p-value: 4.347025911450828e-05


* Here, we can see that the p-value is lesser than the alpha value at 90% confidence (0.1)
* Thus we reject the null hypothesis i.e. we cannot conclude the weight distribution of manual transmission cars is same as the weight distribution of automatic transmission cars